File: 2019/day2/day2.go
package main

import (
	aoc "2019"
	VM "2019/vm"
	"os"
)

func part1(input string) int {
	vm := &VM.VM{}
	vm.Init(VM.MemoryFromProgram(input))
	vm.Memory[1] = 12
	vm.Memory[2] = 2

	vm.Run()

	return vm.Memory[0]
}

func part2(input string) int {
	initialMemory := VM.MemoryFromProgram(input)

	for a := 0; a < 100; a++ {
		for b := 0; b < 100; b++ {
			vm := &VM.VM{}
			vm.Init(initialMemory)
			vm.Memory[1] = a
			vm.Memory[2] = b

			vm.Run()
			if vm.Memory[0] == 19690720 {
				return 100*a + b
			}
		}
	}

	return -1
}

func main() {
	content, _ := os.ReadFile("day2/input.txt")

	aoc.Runner(func() (int, int) {
		a := part1(string(content))
		b := part2(string(content))

		return a, b
	}, "Day 2")
}

File: 2024/lib/day2/day2.ex
defmodule SafeReports do
  @moduledoc "Day 2"

  defp parse_input(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(fn t ->
        String.split(t, " ", trim: true)
        |> Enum.map(&String.to_integer/1)
      end)
  end

  def part1(input) do
    parse_input(input) |>
      Enum.map(fn row ->
        safe?(row)
      end)
      |> Enum.count(& &1)
  end

  defp valid?(a, b, prev_diff) do
    next_diff = a - b
    abs(next_diff) <= 3 && (prev_diff == nil || (next_diff * prev_diff) > 0)
  end

  defp safe?([first | row]) do
    {_, _, valid} = row |> Enum.reduce({first, nil, true}, fn
        _, {_, _, false} -> {0, 0, false}
        v, {prev, prev_diff, _} -> {v, v - prev, valid?(v, prev, prev_diff)}
      end)
    valid
  end

  def part2(input) do
    parse_input(input) |>
      Enum.map(fn row ->
        [row | Enum.map(0..(length(row) - 1), &List.delete_at(row, &1))]
        |> Enum.any?(&safe?/1)
      end)
      |> Enum.count(& &1)
  end
end

{:ok, contents} = File.read("lib/day2/input.txt")
IO.puts(SafeReports.part1(contents))
IO.puts(SafeReports.part2(contents))


Benchee.run(%{
              day2_part1: fn -> contents |> SafeReports.part1() end,
              day2_part2: fn -> contents |> SafeReports.part2() end,
              day2_total: fn ->
                SafeReports.part1(contents)
                SafeReports.part2(contents)
              end}, warmup: 2,
time: 3)
